helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterDataProvider.java

```diff
@@ -0,0 +1,146 @@
+        bestPossibleAssignment);
+  }
+
+  /**
+   * Find all the replications that need to be reallocated.
+   *
+   * @param dataProvider           The cluster status cache that contains the current cluster status.
+   * @param resourceMap            All the valid resources that are managed by the rebalancer.
+   * @param bestPossibleAssignment The persisted Best Possible State.
+   * @param clusterChanges         All the cluster changes that happened after the previous rebalance.
+   * @return A set of assignable replications that need reallocation.
+   */
+  private static Set<AssignableReplica> generateAssignableReplicas(
+      ResourceControllerDataProvider dataProvider, Map<String, Resource> resourceMap,
+      int instanceCount, Map<String, IdealState> bestPossibleAssignment,
+      Map<ClusterDataDetector.ChangeType, Set<String>> clusterChanges) {
+    Set<AssignableReplica> totalReplicas = new HashSet<>();
+
+    for (String resourceName : resourceMap.keySet()) {
+      ResourceConfig config = dataProvider.getResourceConfig(resourceName);
+      IdealState is = dataProvider.getIdealState(resourceName);
+      if (is == null) {
+        throw new HelixException(
+            "Cannot find the resource ideal state for resource: " + resourceName);
+      }
+      String defName = is.getStateModelDefRef();
+      StateModelDefinition def = dataProvider.getStateModelDef(defName);
+      if (def == null) {
+        throw new IllegalArgumentException(String
+            .format("Cannot fine state model definition %s for resource %s.",
+                is.getStateModelDefRef(), resourceName));
+      }
+
+      Map<String, Integer> stateCountMap =
+          def.getStateCountMap(instanceCount, is.getReplicaCount(instanceCount));
+
+      for (String partition : is.getPartitionSet()) {
+        for (Map.Entry<String, Integer> entry : stateCountMap.entrySet()) {
+          String state = entry.getKey();
+          for (int i = 0; i < entry.getValue(); i++) {
+            totalReplicas.add(new AssignableReplica(config, partition, state,
+                def.getStatePriorityMap().get(state)));
+          }
+        }
+      }
+    }
+    return totalReplicas;
+  }
+
+  private static void addNodeAssignmentsToFaultZone(AssignableNode node,
+      Map<String, Map<String, Set<String>>> assignmentForFaultZoneMap) {
+    String faultZoneId = node.getFaultZone();
+    if (!assignmentForFaultZoneMap.containsKey(faultZoneId)) {
+      assignmentForFaultZoneMap.put(faultZoneId, new HashMap<>());
+    }
+    for (Map.Entry<String, Set<String>> entry : node.getCurrentAssignmentsMap().entrySet()) {
+      String resourceName = entry.getKey();
+      if (!assignmentForFaultZoneMap.get(faultZoneId).containsKey(resourceName)) {
+        assignmentForFaultZoneMap.get(faultZoneId).put(resourceName, new HashSet<>());
+      }
+      assignmentForFaultZoneMap.get(faultZoneId).get(resourceName).addAll(entry.getValue());
```

Comment:
`computeIfAbsent` since you've use a lot of Java 8? I think it is fine if you prefer to do in this way. :)